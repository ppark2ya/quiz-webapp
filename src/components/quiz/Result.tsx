import { useHistory, useLocation } from 'react-router-dom';
import { Location } from 'history';
import { IQuizItem } from 'hooks/useQuiz';
import styled from 'styled-components';
import { useCallback, useEffect, useMemo } from 'react';
import Button from 'components/common/Button';
import { ReactComponent as Retry } from 'assets/retry.svg';
import { ReactComponent as Note } from 'assets/check-note.svg';
import PieChart from 'components/common/PieChart';

const Container = styled.div`
  height: 100vh;
  padding: 5vw;
  display: flex;
  flex-direction: column;
  justify-content: flex-end;

  .title {
    font-size: 2.5rem;
    position: fixed;
    top: 4vh;
    left: 0;
    width: 100vw;
    text-align: center;
  }
  .result {
    font-size: 1.2rem;
    padding: 0.3vh 0;
  }
`;

const StyledButton = styled(Button)`
  display: flex;
  justify-content: center;
  margin: 1vh 0;
  svg {
    margin-right: 2vw;
  }
`;

interface IResultProps {
  progressTime: number;
  quizItems: IQuizItem[];
  quizCount: number;
}

function Result() {
  const location: Location<{ results: IResultProps }> = useLocation();
  const history = useHistory();

  const onRetryQuiz = useCallback(() => {
    if (location.state?.results) {
      history.push({
        pathname: '/quiz',
        state: {
          results: location.state.results.quizItems.map((quiz) => ({
            ...quiz,
            userAnswer: undefined,
          })),
        },
      });
    }
  }, [location.state?.results]);

  const onShowWrongNote = useCallback(() => {
    if (location.state?.results) {
      history.push({
        pathname: '/quiz',
        state: {
          results: location.state.results.quizItems,
        },
      });
    }
  }, [location.state?.results]);

  const memoizedComponent = useMemo(() => {
    if (location.state?.results) {
      const { progressTime, quizItems } = location.state.results;
      const { solved, failed } = quizItems?.reduce(
        (prev, quiz) =>
          quiz.userAnswer === quiz.correct_answer
            ? {
                ...prev,
                solved: prev.solved + 1,
              }
            : {
                ...prev,
                failed: prev.failed + 1,
              },
        {
          solved: 0,
          failed: 0,
        },
      );

      return (
        <>
          <PieChart
            style={{ height: '50%', marginTop: '-10vh' }}
            data={[
              {
                id: 'solved',
                label: 'solved',
                value: solved,
                color: 'hsl(60, 70%, 50%)',
              },
              {
                id: 'failed',
                label: 'failed',
                value: failed,
                color: 'hsl(92, 70%, 50%)',
              },
            ]}
            fill={[
              {
                match: {
                  id: 'solved',
                },
                id: 'lines',
              },
              {
                match: {
                  id: 'failed',
                },
                id: 'lines',
              },
            ]}
          />
          <div className="title">Result</div>
          <div className="result">{`Time taken to solve the quiz: ${progressTime}`}</div>
          <div className="result">{`Solved: ${solved}`}</div>
          <div className="result">{`Failed: ${failed}`}</div>
          <StyledButton onClick={onRetryQuiz}>
            <Retry fill="white" />
            Try this again
          </StyledButton>
          <StyledButton onClick={onShowWrongNote}>
            <Note fill="white" />
            Wrong Note
          </StyledButton>
        </>
      );
    } else {
      return <div>The wrong approach.</div>;
    }
  }, [location.state?.results]);

  useEffect(() => {
    if (location.state?.results === undefined) {
      history.push('/');
    }
  }, [location.state?.results]);

  return <Container>{memoizedComponent}</Container>;
}

export default Result;
