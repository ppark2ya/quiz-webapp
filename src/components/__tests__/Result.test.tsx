import { render, screen } from '@testing-library/react';
import Result from '../quiz/Result';
import { ThemeProvider } from 'styled-components';
import { theme } from 'style/theme';
import 'jest-styled-components';
import { useLocation } from 'react-router-dom';

const mockHistoryPush = jest.fn();
jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useLocation: () => ({
    pathname: 'localhost:3000/result',
    state: {
      results: {
        progressTime: 19,
        quizItems: [
          {
            category: 'General Knowledge',
            correct_answer: 'Ed Sheeran - I See Fire',
            difficulty: 'hard',
            incorrect_answers: [
              'Marvin Gaye - Sexual Healing',
              'Coldplay - Midnight',
              'a-ha - Take On Me',
            ],
            question:
              'Electronic music producer Kygo&#039;s popularity skyrocketed after a certain remix. Which song did he remix?',
            type: 'multiple',
            allAnswers: [
              'Ed Sheeran - I See Fire',
              'Marvin Gaye - Sexual Healing',
              'Coldplay - Midnight',
              'a-ha - Take On Me',
            ],
            id: 0,
            useAnswer: 'Marvin Gaye - Sexual Healing',
          },
          {
            category: 'programming Knowledge',
            correct_answer: 'C',
            difficulty: 'easy',
            incorrect_answers: ['python', 'java', 'ruby'],
            question: 'Which of the following languages ​​is the oldest?',
            type: 'multiple',
            allAnswers: ['C', 'python', 'java', 'ruby'],
            id: 1,
            useAnswer: 'C',
          },
        ],
      },
    },
  }),
  useHistory: () => ({
    push: mockHistoryPush,
  }),
}));

test('<Result/>', () => {
  const { container } = render(
    <ThemeProvider theme={theme}>
      <Result />
    </ThemeProvider>,
  );
  const location = useLocation<{ results: any }>();
  console.log(location.state);
  expect(container.firstChild).toMatchSnapshot();

  const timeElement = screen.getByText(/0min 19secs/);
  expect(timeElement).toBeInTheDocument();

  const solvedElement = screen.getByText(/1/);
  expect(solvedElement).toBeInTheDocument();

  const failedElement = screen.getByText(/1/);
  expect(failedElement).toBeInTheDocument();
});
