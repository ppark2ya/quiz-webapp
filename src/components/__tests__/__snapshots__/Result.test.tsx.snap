// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Result/> 1`] = `
.c1 {
  border-radius: 10px;
  background-color: #DA291C;
  color: white;
  font-weight: bold;
  padding: 15px 15vw;
  font-size: 1.2rem;
}

.c0 {
  height: 100vh;
  padding: 5vw;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-box-pack: end;
  -webkit-justify-content: flex-end;
  -ms-flex-pack: end;
  justify-content: flex-end;
}

.c0 .title {
  font-size: 2.5rem;
  position: fixed;
  top: 4vh;
  left: 0;
  width: 100vw;
  text-align: center;
}

.c0 .result {
  font-size: 1.2rem;
  padding: 0.3vh 0;
}

.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  margin: 1vh 0;
}

.c2 svg {
  margin-right: 2vw;
}

<div
  class="c0"
>
  <div
    style="height: 50%; margin-top: -10vh;"
  >
    <div
      style="width: 100%; height: 100%;"
    />
  </div>
  <div
    class="title"
  >
    Result
  </div>
  <div
    class="result"
  >
    Elapsed Time: 0min 19secs
  </div>
  <div
    class="result"
  >
    Solved: 0
  </div>
  <div
    class="result"
  >
    Failed: 2
  </div>
  <button
    class="c1 c2"
    type="submit"
  >
    <svg
      fill="white"
    >
      retry.svg
    </svg>
    Try this again
  </button>
  <button
    class="c1 c2"
    type="submit"
  >
    <svg
      fill="white"
    >
      check-note.svg
    </svg>
    Wrong Note
  </button>
</div>
`;
